src/memory.rs

```diff
@@ -0,0 +1,153 @@
+//! Utilities for converting Kubernetes quantities to Java heap settings.
+
+use k8s_openapi::apimachinery::pkg::api::resource::Quantity;
+
+use crate::error::{Error, OperatorResult};
+use std::str::FromStr;
+
+#[derive(Clone, Debug, PartialEq, PartialOrd)]
+pub enum BinaryMultiple {
+    Kibi,
+    Mebi,
+    Gibi,
+    Tebi,
+    Pebi,
+    Exbi,
+}
+
+impl BinaryMultiple {
+    pub fn to_legacy(&self) -> String {
+        match self {
+            BinaryMultiple::Kibi => "k".to_string(),
+            BinaryMultiple::Mebi => "m".to_string(),
+            BinaryMultiple::Gibi => "g".to_string(),
+            BinaryMultiple::Tebi => "t".to_string(),
+            BinaryMultiple::Pebi => "p".to_string(),
+            BinaryMultiple::Exbi => "e".to_string(),
+        }
+    }
+
+    pub fn upscale(&self) -> Self {
+        match self {
+            BinaryMultiple::Kibi => BinaryMultiple::Kibi,
+            BinaryMultiple::Mebi => BinaryMultiple::Kibi,
+            BinaryMultiple::Gibi => BinaryMultiple::Mebi,
+            BinaryMultiple::Tebi => BinaryMultiple::Gibi,
+            BinaryMultiple::Pebi => BinaryMultiple::Tebi,
+            BinaryMultiple::Exbi => BinaryMultiple::Pebi,
+        }
+    }
+}
+
+impl FromStr for BinaryMultiple {
+    type Err = Error;
+
+    fn from_str(q: &str) -> OperatorResult<BinaryMultiple> {
+        let lq = q.to_lowercase();
+        match lq.as_str() {
+            "ki" | "kib" => Ok(BinaryMultiple::Kibi),
+            "mi" | "mib" => Ok(BinaryMultiple::Mebi),
+            "gi" | "gib" => Ok(BinaryMultiple::Gibi),
+            "ti" | "tib" => Ok(BinaryMultiple::Tebi),
+            "pi" | "pib" => Ok(BinaryMultiple::Pebi),
+            "ei" | "eib" => Ok(BinaryMultiple::Exbi),
+            _ => Err(Error::InvalidQuantityUnit {
+                value: q.to_string(),
+            }),
+        }
```

Comment:
Capitalization matters, for example `mi` should be milli (or fail loudly). K8s quantities also generally leave the unit itself (`B` here) out.

Reply:
milli comes from https://kubernetes.io/docs/tasks/configure-pod-container/assign-cpu-resource/.. looks like the unit for that is `m` rather than `mi`. Still applies though that `1mB` is a very different thing from `1MB`.